File: builder/java/BuilderBasic.java
class Product {
  private String describe;
  private String name;

  void withNameDes(String name, String describe) {
    this.name = name;
    this.describe = describe;
  }

  @Override
  public String toString() {
    return "Product{" +
        "describe='" + describe + '\'' +
        ", name='" + name + '\'' +
        '}';
  }
}

abstract class Builder {
  Product product = null;
  abstract void buildPart();
  Product getResult(){
    System.out.println(product);
    return product;
  }
}

class ConcreteBuilder extends Builder {
  @Override
  void buildPart() {
    product = new Product();
    product.withNameDes("product", "Product instance has been created");
  }
}

class Director {
  Builder builder;
  Director(Builder builder) {
    this.builder = builder;
  }

  void construct() {
    builder.buildPart();
  }
}

public class BuilderBasic {
  public static void main(String[] args) {
    Builder builder = new ConcreteBuilder();
    Director director = new Director(builder);
    director.construct();
    builder.getResult();
  }
}

File: builder/python/BuilderBasic.py
#!/usr/bin/env python
# coding:utf8


class Product(object):
    def withNameDes(self, **kwargs):
        self._name = kwargs.get("name")
        self._des = kwargs.get("describe")

    def __repr__(self, *args, **kwargs):
        return "name: " + self._name + "\ndescribe: " + self._des


class Builder(object):
    def buildePart(self):
        pass

    def getResult(self):
        pass


class ConcreteBuilder(Builder):
    def buildePart(self):
        self._product = Product()
        self._product.withNameDes(
            name="product", describe="Product instance has been created"
        )

    def getResult(self):
        print(self._product)
        return self._product


if __name__ == "__main__":
    builder = ConcreteBuilder()
    builder.buildePart()
    builder.getResult()
